server/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java

```diff
@@ -234,16 +235,27 @@ static SecureSettings loadSecureSettings(Environment initialEnv) throws Bootstra
             throw new BootstrapException(e);
         }
 
+        char[] password;
+        if (keystore != null && keystore.hasPassword()) {
+            password = Terminal.DEFAULT.readSecret("Elasticsearch password? ");
+        } else {
+            // TODO[wrb]: is there a case w/o stdin?
+            password = new char[0];
+        }
+
         try {
             if (keystore == null) {
                 final KeyStoreWrapper keyStoreWrapper = KeyStoreWrapper.create();
                 keyStoreWrapper.save(initialEnv.configFile(), new char[0]);
                 return keyStoreWrapper;
             } else {
-                keystore.decrypt(new char[0] /* TODO: read password from stdin */);
-                KeyStoreWrapper.upgrade(keystore, initialEnv.configFile(), new char[0]);
+                keystore.decrypt(password);
+                KeyStoreWrapper.upgrade(keystore, initialEnv.configFile(), password);
             }
         } catch (Exception e) {
+            if (password.length != 0) {
```

Comment:
`decrypt` would throw a `SecurityException` with the message "Provided keystore password was incorrect" if the password was incorrect so you can catch that specifically and wrap it in BootstrapException instead of basing the check on whether the user provided a password or not

Reply:
I think I was seeing something different earlier in development, but now I'm seeing what you describe. Thanks!

Reply:
It could be that I haven't merged that particular change when you originally wrote the code :)